File: frontend/src/utils/dateRange.ts
/**
 * Date range utilities for formatting, validation, and UUID generation.
 *
 * Mirrors the backend normalization so the frontend can build the exact
 * payload stored in property_value_scalar.value_text.
 */

import { generateDateUuid, type ParsedDate } from './dateParser';

export type DateRangeGranularity = 'day' | 'month' | 'year';

export interface DateRangeValue {
  start: string;          // ISO date YYYY-MM-DD
  end: string;
  granularity: DateRangeGranularity;
  start_uuid: string;
  end_uuid: string;
}

const MONTH_NAMES = [
  '', 'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December',
];

function pad(n: number, len: number): string {
  return String(n).padStart(len, '0');
}

function parseIsoDate(value: string): Date {
  const d = new Date(value);
  if (Number.isNaN(d.getTime())) {
    throw new Error(`Invalid date: ${value}`);
  }
  // Use local midnight so the displayed date matches the stored string.
  const [year, month, day] = value.split('-').map((p) => parseInt(p, 10));
  return new Date(year, month - 1, day);
}

function canonicalDates(
  start: Date,
  end: Date,
  granularity: DateRangeGranularity,
): { start: Date; end: Date } {
  if (granularity === 'day') {
    return { start, end };
  }

  if (granularity === 'month') {
    const startCanonical = new Date(start.getFullYear(), start.getMonth(), 1);
    const endCanonical = new Date(end.getFullYear(), end.getMonth() + 1, 0);
    return { start: startCanonical, end: endCanonical };
  }

  // year
  return {
    start: new Date(start.getFullYear(), 0, 1),
    end: new Date(end.getFullYear(), 11, 31),
  };
}

function isoString(d: Date): string {
  return `${d.getFullYear()}-${pad(d.getMonth() + 1, 2)}-${pad(d.getDate(), 2)}`;
}

function uuidForDate(d: Date, granularity: DateRangeGranularity): string {
  const parsed: ParsedDate = {
    type: granularity,
    year: d.getFullYear(),
    month: granularity !== 'year' ? d.getMonth() + 1 : undefined,
    day: granularity === 'day' ? d.getDate() : undefined,
    label: '',
  };
  return generateDateUuid(parsed);
}

/**
 * Validate and normalize a date range payload.
 */
export function normalizeDateRange(
  start: string,
  end: string,
  granularity: DateRangeGranularity,
): DateRangeValue {
  if (!['day', 'month', 'year'].includes(granularity)) {
    throw new Error(`Invalid granularity: ${granularity}`);
  }

  const startDate = parseIsoDate(start);
  const endDate = parseIsoDate(end);

  if (startDate.getTime() > endDate.getTime()) {
    throw new Error('Start date must be before or equal to end date');
  }

  const { start: startCanonical, end: endCanonical } = canonicalDates(startDate, endDate, granularity);

  return {
    start: isoString(startCanonical),
    end: isoString(endCanonical),
    granularity,
    start_uuid: uuidForDate(startCanonical, granularity),
    end_uuid: uuidForDate(endCanonical, granularity),
  };
}

/**
 * Format a date range for display.
 */
export function formatDateRange(value: DateRangeValue | null): string {
  if (!value) return '';
  const { start, end, granularity } = value;
  const startDate = parseIsoDate(start);
  const endDate = parseIsoDate(end);

  if (granularity === 'day') {
    if (start === end) {
      return formatDay(startDate);
    }
    return `${formatDay(startDate)} – ${formatDay(endDate)}`;
  }

  if (granularity === 'month') {
    if (startDate.getFullYear() === endDate.getFullYear() && startDate.getMonth() === endDate.getMonth()) {
      return formatMonth(startDate);
    }
    return `${formatMonth(startDate)} – ${formatMonth(endDate)}`;
  }

  if (startDate.getFullYear() === endDate.getFullYear()) {
    return String(startDate.getFullYear());
  }
  return `${startDate.getFullYear()} – ${endDate.getFullYear()}`;
}

function formatDay(d: Date): string {
  return `${MONTH_NAMES[d.getMonth() + 1]} ${d.getDate()}, ${d.getFullYear()}`;
}

function formatMonth(d: Date): string {
  return `${MONTH_NAMES[d.getMonth() + 1]} ${d.getFullYear()}`;
}

/**
 * Parse a target date UUID (day/month/year journal page) into a canonical ISO date.
 */
export function dateUuidToIso(uuid: string): string | null {
  if (!uuid || uuid.length !== 36) return null;

  // Day UUID: 00000000-0000-0000-00dd-YYYYMMDD0000
  if (uuid.startsWith('00000000-0000-0000-00dd-')) {
    const data = uuid.slice(-12, -4);
    if (/^\d{8}$/.test(data)) {
      const y = data.slice(0, 4);
      const m = data.slice(4, 6);
      const d = data.slice(6, 8);
      return `${y}-${m}-${d}`;
    }
  }

  // Month UUID: 00000000-0000-0000-00aa-YYYYMM000000
  if (uuid.startsWith('00000000-0000-0000-00aa-')) {
    const data = uuid.slice(-12, -6);
    if (/^\d{6}$/.test(data)) {
      const y = data.slice(0, 4);
      const m = data.slice(4, 6);
      return `${y}-${m}-01`;
    }
  }

  // Year UUID: 00000000-0000-0000-00bb-YYYY00000000
  if (uuid.startsWith('00000000-0000-0000-00bb-')) {
    const data = uuid.slice(-12, -8);
    if (/^\d{4}$/.test(data)) {
      return `${data}-01-01`;
    }
  }

  return null;
}

/**
 * Return the granularity of a date UUID, or null if it is not a date UUID.
 */
export function dateUuidGranularity(uuid: string): DateRangeGranularity | null {
  if (!uuid || uuid.length !== 36) return null;
  if (uuid.startsWith('00000000-0000-0000-00dd-')) return 'day';
  if (uuid.startsWith('00000000-0000-0000-00aa-')) return 'month';
  if (uuid.startsWith('00000000-0000-0000-00bb-')) return 'year';
  return null;
}

/**
 * Build a DateRangeValue from raw inputs, suitable for API payloads.
 */
export function buildDateRangeValue(
  start: string,
  end: string,
  granularity: DateRangeGranularity,
): DateRangeValue {
  return normalizeDateRange(start, end, granularity);
}

File: app/utils/date_range.py
"""Date range normalization and validation utilities.

Date ranges are stored as JSON in property_value_scalar.value_text. This module
produces the canonical payload and the deterministic journal-page UUIDs needed
for inline links and QueryAST containment checks.
"""

from __future__ import annotations

import json
from datetime import date, datetime
from enum import StrEnum
from typing import Any

from app.domain.entities.constants import (
    generate_day_uuid,
    generate_month_uuid,
    generate_year_uuid,
    parse_date_uuid,
)


class DateRangeGranularity(StrEnum):
    """Granularity of a date range."""

    DAY = "day"
    MONTH = "month"
    YEAR = "year"


def _parse_iso_date(value: str | date | datetime) -> date:
    """Parse an ISO date string or date/datetime object into a date."""
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    if isinstance(value, str):
        value = value.strip()
        if not value:
            raise ValueError("Date value is empty")
        # Allow "YYYY-MM-DD" or datetime strings; take the date portion.
        if "T" in value:
            value = value.split("T")[0]
        return date.fromisoformat(value)
    raise ValueError(f"Invalid date value: {value!r}")


def _canonical_dates(start: date, end: date, granularity: DateRangeGranularity) -> tuple[date, date]:
    """Return canonical start/end dates for the requested granularity."""
    if granularity == DateRangeGranularity.DAY:
        return start, end

    if granularity == DateRangeGranularity.MONTH:
        start_canonical = date(start.year, start.month, 1)
        # Last day of end month: first day of next month minus one day.
        first_next = date(end.year + 1, 1, 1) if end.month == 12 else date(end.year, end.month + 1, 1)
        end_canonical = date.fromordinal(first_next.toordinal() - 1)
        return start_canonical, end_canonical

    # YEAR
    start_canonical = date(start.year, 1, 1)
    end_canonical = date(end.year, 12, 31)
    return start_canonical, end_canonical


def _uuid_for_date(d: date, granularity: DateRangeGranularity) -> str:
    """Generate the deterministic journal-page UUID for a date at a granularity."""
    if granularity == DateRangeGranularity.DAY:
        return generate_day_uuid(d)
    if granularity == DateRangeGranularity.MONTH:
        return generate_month_uuid(d.year, d.month)
    return generate_year_uuid(d.year)


def date_uuid_to_date(uuid: str) -> date:
    """Convert a deterministic journal-page UUID into its canonical date.

    Day UUIDs map to the exact day. Month and year UUIDs map to the first day
    of that month/year so they can be compared uniformly with date ranges.
    """
    info = parse_date_uuid(uuid)
    if info is None:
        raise ValueError(f"Not a date UUID: {uuid}")

    if info["type"] == "day":
        return date(info["year"], info["month"], info["day"])
    if info["type"] == "month":
        return date(info["year"], info["month"], 1)
    return date(info["year"], 1, 1)


def date_uuid_granularity(uuid: str) -> DateRangeGranularity:
    """Return the granularity of a date UUID."""
    info = parse_date_uuid(uuid)
    if info is None:
        raise ValueError(f"Not a date UUID: {uuid}")
    if info["type"] == "day":
        return DateRangeGranularity.DAY
    if info["type"] == "month":
        return DateRangeGranularity.MONTH
    return DateRangeGranularity.YEAR


def normalize_date_range(
    start: str | date | datetime,
    end: str | date | datetime,
    granularity: str,
) -> dict[str, Any]:
    """Normalize and validate a date range payload.

    Args:
        start: Start date (ISO string or date object).
        end: End date (ISO string or date object).
        granularity: "day", "month", or "year".

    Returns:
        Dict with canonical start/end ISO dates, granularity, and UUIDs.

    Raises:
        ValueError: If inputs are invalid or start > end.
    """
    try:
        g = DateRangeGranularity(granularity)
    except ValueError as exc:
        raise ValueError(f"Invalid granularity: {granularity!r}") from exc

    start_date = _parse_iso_date(start)
    end_date = _parse_iso_date(end)

    if start_date > end_date:
        raise ValueError("Start date must be before or equal to end date")

    start_canonical, end_canonical = _canonical_dates(start_date, end_date, g)

    return {
        "start": start_canonical.isoformat(),
        "end": end_canonical.isoformat(),
        "granularity": g.value,
        "start_uuid": _uuid_for_date(start_canonical, g),
        "end_uuid": _uuid_for_date(end_canonical, g),
    }


def normalize_date_range_value(value: Any) -> str:
    """Normalize a date range value and return its JSON string.

    Accepts:
      - A JSON string representing the payload.
      - A dict with start, end, granularity.

    Returns:
        Canonical JSON string for storage in property_value_scalar.value_text.
    """
    if isinstance(value, str):
        value = value.strip()
        if not value:
            raise ValueError("date_range value is empty")
        try:
            parsed = json.loads(value)
        except json.JSONDecodeError as exc:
            raise ValueError("date_range value is not valid JSON") from exc
        value = parsed

    if not isinstance(value, dict):
        raise ValueError("date_range value must be an object")

    start = value.get("start")
    end = value.get("end")
    granularity = value.get("granularity")
    if start is None or end is None or granularity is None:
        raise ValueError("date_range value must include start, end, and granularity")

    normalized = normalize_date_range(start, end, granularity)
    return json.dumps(normalized, separators=(",", ":"), sort_keys=True)
